lib/oper.g

```diff
@@ -1922,6 +1917,96 @@ BIND_GLOBAL( "FLUSH_ALL_METHOD_CACHES", function()
 end);
 
 fi;
+#############################################################################
+##
+#F RECALCULATE_ALL_METHOD_RANKS() . . reorder methods after new implications
+##
+## Installing new implications (including hidden implications) can change the
+## rank of existing filters, and so of existing methods for operations.
+##
+## This function recalculates all such ranks and adjusts the method ordering
+## where needed. If the ordering changes, the relevant caches are flushed.
+##
+## If PRINT_REORDERED_METHODS is true, it prints some diagnostics (this is a
+## bit too low-level for Info).
+##
+##
+
+Unbind(RECALCULATE_ALL_METHOD_RANKS);
+
+PRINT_REORDERED_METHODS := false;
+
+BIND_GLOBAL( "RECALCULATE_ALL_METHOD_RANKS", function()
+    local  oper, n, changed, meths, nmethods, i, base, rank, j, req, 
+           req2, k, l;
+    
+    for oper in OPERATIONS do
+        if IS_CONSTRUCTOR(oper) then
+            # FIXME do something for constructors
+            continue;
+        fi;
+        
+        for n in [0..6] do
+            changed := false;                
+            meths := METHODS_OPERATION(oper, n);
+            nmethods := LENGTH(meths)/(BASE_SIZE_METHODS_OPER_ENTRY+n);                
+            for i in [1 ..nmethods] do
+                base := (i-1)*(BASE_SIZE_METHODS_OPER_ENTRY+n);
+                # data for this method is meths{[base+1..base+BASE_SIZE_METHODS_OPER_ENTRY + n]}
+                rank := meths[base+6+n];                    
+                if IS_FUNCTION(rank) then
+                    rank := rank();
+                fi;
+                for j in [1..n] do
+                    req := meths[base+1+j];
+                    rank := rank+RankFilter(WITH_IMPS_FLAGS(req));
+                od;
+                if rank = meths[base+n+3] then
+                    continue;
+                fi;
+                meths[base+n+3] := rank;
+                if i = 1 or rank <= meths[base-3] then
```

Comment:
Why -3? Wait, is that trying to access the previous method's entry? But that could break if `BASE_SIZE_METHODS_OPER_ENTRY` changes again, no? It's also not immediately clear which "field" `base-3` refers to. While this makes it clear: `meths[base+n+3 - BASE_SIZE_METHODS_OPER_ENTRY]` (I'd still ad a comment to explain what's going on, though?)

Reply:
Will do